### ISSUE TITLE
Improve Testability of Map Click Events

### ISSUE THREAD

**Issue by gchallen** [osmdroid/osmdroid#1859]
## Issue Type

[ x ] Improvement

## Description and/or steps/code to reproduce the problem

I'm using `osmdroid` as part of an Android programming project for an introductory programming course (https://www.cs124.org/). In order to evaluate student submissions efficiently in a trusted environment, we make heavy use of Robolectric to perform UI testing without needing to perform full system emulation. (Too slow.)

Overall I've been able to test `osmdroid` quite effectively using Robolectric. But I've encountered a pain point around delivering click events to the map. For example, I'd like to test to ensure that a long press on the `MapView` launches a new activity providing the user with a chance to add information to the map database about that position. Or that a click on the `MapView` that intersects an existing marker causes its popup to open properly.

I was eventually able to accomplish this, but it required using reflection to open several private fields on the `MapView` class:

```java
Field mGestureDetector = MapView.class.getDeclaredField("mGestureDetector");
mGestureDetector.setAccessible(true);
GestureDetector detector = (GestureDetector) mGestureDetector.get(mapView);

Field mDoubleTapListener = GestureDetector.class.getDeclaredField("mDoubleTapListener");
mDoubleTapListener.setAccessible(true);
doubleTapListener = (GestureDetector.OnDoubleTapListener) mDoubleTapListener.get(detector);

Field mListener = GestureDetector.class.getDeclaredField("mListener");
mListener.setAccessible(true);
gestureListener = (GestureDetector.OnGestureListener) mListener.get(detector);
```

Once this is done, click events can be delivered by creating a `MotionEvent` and passing it to `doubleTapListener.onSingleTapConfirmed`, and long press events can be delivered by passing the `MotionEvent` to `gestureListener.onLongPress`. But obviously using reflection here to access a private API isn't a great idea.

The simplest fix would be to just make these fields publicly accessible. Another option would be to provide a testing API allowing these events to be delivered to the `MapView`, and potentially handle the geo-coordinates to pixel-coordinates transformation, another somewhat non-obvious bit that would make testing a bit simpler, since usually geo-coordinates are more meaningful.

Overall great work on this project! I'm really excited that my students will be able to develop a mapping application using an open-source library.

**Comment by eclectice** [comment]
You have to implement the **MapEventsReceiver*** interface in your activity or fragment to expose the single tap and long press events upon the MapView.

https://github.com/osmdroid/osmdroid/blob/dbb20901cac30b28a6a9349da097c22d9a962311/osmdroid-android/src/main/java/org/osmdroid/events/MapEventsReceiver.java#L12

```java
/**
 * Interface for objects that need to handle map events thrown by a MapEventsOverlay.
 *
 * @author M.Kergall
 * @see MapEventsOverlay
 */
public interface MapEventsReceiver {

    /**
     * @param p the position where the event occurred.
     * @return true if the event has been "consumed" and should not be handled by other objects.
     */
    boolean singleTapConfirmedHelper(GeoPoint p);

    /**
     * @param p the position where the event occurred.
     * @return true if the event has been "consumed" and should not be handled by other objects.
     */
    boolean longPressHelper(GeoPoint p);
}
```

**Comment by gchallen** [comment]
Right, we're doing this in the app. The suggestion is on how to deliver such events during Robolectric testing, not how to handle them.

**Comment by eclectice** [comment]
What about extending the **Overlay** class to give you direct access to its ***onSingleTapConfirmed()*** and ***onLongPress()*** methods, which work similarly to **MapEventsOverlay** but are now adapted for the Robolectric tests?

https://github.com/osmdroid/osmdroid/blob/dbb20901cac30b28a6a9349da097c22d9a962311/osmdroid-android/src/main/java/org/osmdroid/views/overlay/MapEventsOverlay.java#L40

https://github.com/osmdroid/osmdroid/blob/dbb20901cac30b28a6a9349da097c22d9a962311/osmdroid-android/src/main/java/org/osmdroid/views/overlay/MapEventsOverlay.java#L47

**Comment by gchallen** [comment]
The problem with extending `Overlay` is then you need to find the right overlay to deliver the event to during testing. In our case, we're using the provided `Marker` class, which requires one `Overlay` per `Marker`, so now we have to loop and deliver the event to every `Overlay`, or duplicate the hit detection logic that's already present on the `MapView` class.

I think ideally the click and touch events are delivered to the `MapView`, and then it can determine (in the usual way) which `Overlay` to deliver them to. To me this interface is more intuitive ("the user long-pressed at this lat-lon"), simpler, and allows you to test things like the impact of overlapping markers.

I'm not 100% sure how extension works with Robolectric. I'd have to try it. I suppose you could then create a shadow that extended the `MapView` class and provided the needed functionality. Or maybe that shadow could be provided by `osmdroid`?

The lowest friction way that this works right out of the box is by modifying the existing `MapView` class, but I can understand if you don't want to add a new API only for testing. (Although perhaps there are some use cases for programmatic access to `MapView` gestures.)
